src/OrchardCore.Modules/OrchardCore.AutoSetup/AutoSetupMiddleware.cs

```diff
@@ -0,0 +1,182 @@
+using System;
+using OrchardCore.AutoSetup.Options;
+using OrchardCore.Environment.Shell.Models;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Scope;
+using OrchardCore.Setup.Services;
+
+namespace OrchardCore.AutoSetup
+{
+    using OrchardCore.Abstractions.Setup;
+
+    /// <summary>
+    /// The auto setup middleware.
+    /// </summary>
+    public class AutoSetupMiddleware
+    {
+        /// <summary>
+        /// The next middleware in the execution pipeline.
+        /// </summary>
+        private readonly RequestDelegate _next;
+
+        /// <summary>
+        /// The Shell settings manager.
+        /// </summary>
+        private readonly IShellSettingsManager _shellSettingsManager;
+
+        /// <summary>
+        /// The auto-setup options.
+        /// </summary>
+        private readonly IOptions<AutoSetupOptions> _setupOptions;
+
+        /// <summary>
+        /// The _logger.
+        /// </summary>
+        private readonly ILogger<AutoSetupMiddleware> _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AutoSetupMiddleware"/> class.
+        /// </summary>
+        /// <param name="next"> The next middleware in the execution pipeline. </param>
+        /// <param name="shellSettingsManager">The Shell settings manager</param>
+        /// <param name="setupOptions"> The auto-setup Options. </param>
+        /// <param name="logger"> The logger. </param>
+        public AutoSetupMiddleware(
+            RequestDelegate next,
+            IShellSettingsManager shellSettingsManager,
+            IOptions<AutoSetupOptions> setupOptions,
+            ILogger<AutoSetupMiddleware> logger)
+        {
+            _logger = logger;
+            _next = next;
+            _shellSettingsManager = shellSettingsManager;
+            _setupOptions = setupOptions;
+        }
+
+        /// <summary>
+        /// The middleware auto-setup invoke.
+        /// </summary>
+        /// <param name="httpContext">
+        /// The http context.
+        /// </param>
+        /// <returns>
+        /// The <see cref="Task"/>.
+        /// </returns>
+        public async Task InvokeAsync(HttpContext httpContext)
+        {
+            var currentShellSettings = ShellScope.Context.Settings;
+            var setupService = httpContext.RequestServices.GetRequiredService<ISetupService>();
```

Comment:
Just to say that here we use the `AsyncLocal ShellScope` to retrieve the settings, but not to resolve the service ;)

For me that's okay to use it or not for both, so just for infos the other way would be

           var currentShellSettings = httpContext.RequestServices.GetRequiredService<ShellSettings>();